lib/OpenQA/WebAPI/Controller/Main.pm

```diff
@@ -20,130 +20,38 @@ use warnings;
-                }
-
-                if (   $job->result eq OpenQA::Schema::Result::Jobs::FAILED
-                    || $job->result eq OpenQA::Schema::Result::Jobs::INCOMPLETE)
-                {
-                    $jr{failed}++;
-                    $jr{labeled}++ if $labels{$job->id};
-                    next;
-                }
-                if (grep { $job->result eq $_ } OpenQA::Schema::Result::Jobs::INCOMPLETE_RESULTS) {
-                    next;    # ignore the rest
-                }
-            }
-            if (   $job->state eq OpenQA::Schema::Result::Jobs::CANCELLED
-                || $job->state eq OpenQA::Schema::Result::Jobs::OBSOLETED)
-            {
-                next;        # ignore
-            }
-            my $state = $job->state;
-            if (grep { /$state/ } (OpenQA::Schema::Result::Jobs::PENDING_STATES)) {
-                $jr{inprogress}++;
-                next;
-            }
-            $self->app->log->error("MISSING S:" . $job->state . " R:" . $job->result);
-        }
-        $jr{reviewed_all_passed} = $jr{passed} == $count;
-        $jr{reviewed}            = $jr{failed} > 0 && $jr{labeled} == $jr{failed};
-        $res{$b}                 = \%jr;
-        $max_jobs = $count if ($count > $max_jobs);
-        last if (++$buildnr >= $limit);
-    }
-    $res{_max} = $max_jobs if %res;
-
-    return \%res;
-
-}
+use OpenQA::BuildResults;
 
 sub index {
     my ($self) = @_;
 
-    my $limit_builds = $self->param('limit_builds') // 3;
+    my $limit_builds    = $self->param('limit_builds')    // 3;
+    my $time_limit_days = $self->param('time_limit_days') // 14;
+    $self->app->log->debug("Retrieving results for up to $limit_builds builds up to $time_limit_days days old");
     my $group_params = $self->every_param('group');
-    my @results;
 
+    my @results;
     my $groups = $self->db->resultset('JobGroups')->search({}, {order_by => qw/name/});
+
     while (my $group = $groups->next) {
         next unless (scalar @{$group_params} == 0 || (grep { $group->name =~ /$_/ } @$group_params));
-        my $res = $self->_group_result($group, $limit_builds);
-        if (%$res) {
-            $res->{_group} = $group;
-            push(@results, $res);
-        }
+        my $build_results = OpenQA::BuildResults::compute_build_results($self->app, $group, $limit_builds, $time_limit_days);
+        push(@results, $build_results) if $build_results;
     }
     $self->stash('results', \@results);
 }
 
 sub group_overview {
     my ($self) = @_;
 
-    my $limit_builds = $self->param('limit_builds') // 10;
-    my $only_tagged  = $self->param('only_tagged')  // 0;
-    my $group        = $self->db->resultset('JobGroups')->find($self->param('groupid'));
+    my $limit_builds    = $self->param('limit_builds')    // 10;
+    my $time_limit_days = $self->param('time_limit_days') // 14;
+    $self->app->log->debug("Retrieving results for up to $limit_builds builds up to $time_limit_days days old");
```

Comment:
It depends on whether the default values will always be the same in both cases or might differ in the future. I wasn't sure so I've kept them separate.